File: 0125-valid-palindrome/0125-valid-palindrome.py
# Aug 11, 2026 08:24
# two pointer approach

class Solution:
    def isPalindrome(self, s: str) -> bool:
        # our list of valid characters (alphanumeric)
        alnum = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

        # we create two pointers, one at the start and one at the end of the string
        left = 0
        right = len(s) - 1

        while (left < right):
            if s[left] not in alnum:
                left += 1
            elif s[right] not in alnum:
                right -= 1
            else:
                # we convert them to lowercase first
                if ((s[left]).lower() != (s[right]).lower()):
                    return False
                left += 1
                right -= 1
            
        return True
        

File: 0125-valid-palindrome/0125-valid-palindrome.c
// Aug 11, 2026 08:18
// two pointer approach

bool isPalindrome(char* s) {
    int length = strlen(s);

    // we create two pointers, one starting at the start of the string
    // and one starting at the end of the string
    int left = 0;
    int right = length - 1;

    while (left < right) {
        // we check if either of the current characters are 
        // part of the alphabet. if they aren't, we skip them
        if (!isalnum(s[left])) {
            left++;
        }
        else if (!isalnum(s[right])) {
            right--;
        }
        else {
            // we convert the characters to lowercase
            if (tolower(s[left]) != tolower(s[right])) {
                return false;
            }
            left++;
            right--;
        }
    }

    return true;
}
